extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -0,0 +1,617 @@
+
+  protected final String supervisorRunId;
+  private final TaskStorage taskStorage;
+  private final TaskMaster taskMaster;
+  private final KafkaSupervisorConfig config;
+  private final IndexerMetadataStorageCoordinator indexerMetadataStorageCoordinator;
+  private final KafkaSupervisorSpec spec;
+  private final String dataSource;
+
+  private KafkaConsumer consumer;
+
+  private volatile boolean started = false;
+
+  private final HashMap<Integer, TaskGroup> taskGroups = new HashMap<>();
+  private final ExecutorService exec;
+
+  @Inject
+  public KafkaSupervisor(
+      TaskStorage taskStorage,
+      TaskMaster taskMaster,
+      KafkaSupervisorConfig config,
+      IndexerMetadataStorageCoordinator indexerMetadataStorageCoordinator,
+      KafkaSupervisorSpec spec
+  )
+  {
+    this.taskStorage = taskStorage;
+    this.taskMaster = taskMaster;
+    this.config = config;
+    this.indexerMetadataStorageCoordinator = indexerMetadataStorageCoordinator;
+    this.spec = spec;
+
+    this.dataSource = spec.getDataSchema().getDataSource();
+
+    this.supervisorRunId = getRandomId();
+    this.exec = Execs.singleThreaded(String.format("KafkaSupervisor-%s", dataSource));
+
+    log.info("Initialized with supervisor run ID: [%s]", this.supervisorRunId);
+  }
+
+  @Override
+  public void start()
+  {
+    Preconditions.checkState(!started, "already started");
+    Preconditions.checkState(!exec.isShutdown(), "already stopped");
+
+    try {
+      this.consumer = getKafkaConsumer(this.spec.getIoConfig().getKafkaBrokers());
+      exec.submit(
+          new Runnable()
+          {
+            @Override
+            public void run()
+            {
+              mainLoop();
+            }
+          }
+      );
+    }
+    catch (Exception e) {
+      log.makeAlert(e, "Exception starting KafkaSupervisor[%s]", dataSource)
+         .emit();
+      throw Throwables.propagate(e);
+    }
+
+    started = true;
+    log.info("Started KafkaSupervisor[%s].", dataSource);
+  }
+
+  @Override
+  public void stop()
+  {
+    Preconditions.checkState(started, "not started");
+
+    try {
+      exec.shutdownNow();
+      consumer.close();
+      started = false;
+
+      log.info("KafkaSupervisor[%s] has stopped.", dataSource);
+    }
+    catch (Exception e) {
+      log.makeAlert(e, "Exception stopping KafkaSupervisor[%s]", dataSource)
+         .emit();
+    }
+  }
+
+  @Override
+  public Optional<SupervisorSpec> getSpec()
+  {
+    return Optional.fromNullable((SupervisorSpec) this.spec);
+  }
+
+  private void mainLoop()
+  {
+    try {
+      while (!Thread.currentThread().isInterrupted()) {
+        try {
+          killUnownedKafkaTasks();
+          updatePartitionDataFromKafka();
+          updateTaskMetadata();
+          checkTaskState();
+        }
+        catch (Exception e) {
+          log.makeAlert(e, "Exception in KafkaSupervisor[%s]", dataSource)
+             .emit();
+        }
+
+        Thread.sleep(config.getRunPeriod());
```

Comment:
minor: I think usually "run every X" is easier to reason about than "sleep for X every run". You could get that by remembering the start time of each loop and then doing some math on that.